MobileWallet/Backup/ICloudBackup.swift

```diff
@@ -0,0 +1,342 @@
+            return NSLocalizedString("You have not any wallet backup", comment: "'No any wallet backup' error description")
+        case .unzipError:
+            return NSLocalizedString("Unable to unarchive wallet from iCloud backup", comment: "unarchive wallet error description")
+        case .dbFileNotFound:
+            return NSLocalizedString("Unable to create wallet backup. File not found", comment: "sqlite file not found error description")
+        case .iCloudContainerNotFound:
+            return NSLocalizedString("Unable to create wallet backup. iCloud container not found", comment: "iCloud container not found error description")
+        case .unableCreateBackupFolder:
+            return NSLocalizedString("Unable to create backup folder", comment: "Unable to create backup folder error descroption")
+        case .privateKeyError:
+            return NSLocalizedString("Unable to restore wallet. Private key error", comment: "Unable to restore wallet private key error descroption")
+        }
+    }
+}
+
+protocol ICloudBackupObserver: AnyObject {
+    func didFinishUploadBackup(percent: Double, completed: Bool, error: Error?)
+}
+
+class ICloudBackup: NSObject {
+
+    var query: NSMetadataQuery!
+
+    private let containerIdentifier = "iCloud.com.tari.wallet"
+    private let directory = TariLib.shared.databaseDirectory
+    private let fileName = "Tari-Aurora-Backup"
+    private var observers = NSPointerArray.weakObjects()
+
+    var inProgress: Bool = false
+    private(set) var progressValue: Double = 0.0
+
+    static let shared = ICloudBackup()
+
+    override init() {
+        super.init()
+        initialiseQuery()
+        addNotificationObservers()
+    }
+
+    func initialiseQuery() {
+        query = NSMetadataQuery.init()
+        query.operationQueue = .main
+        query.searchScopes = [NSMetadataQueryUbiquitousDataScope]
+        query.predicate = NSPredicate(format: "%K LIKE '*.zip'", NSMetadataItemFSNameKey)
+    }
+
+    func addObserver(_ observer: ICloudBackupObserver) {
+        observers.addObject(observer)
+    }
+
+    // returns true if backup of current wallet is exist
+    func isBackupExist() -> Bool {
+        let fileManager = FileManager.default
+        guard let backupFolder = TariLib.shared.tariWallet?.publicKey.0?.hex.0 else { return false }
+
+        do {
+            let icloudFolderURL = try iCloudDirectory()
+
+            if let urls = try? fileManager.contentsOfDirectory(at: icloudFolderURL, includingPropertiesForKeys: nil, options: []) {
+                if let _ = urls.first(where: { $0.absoluteString.contains(backupFolder) }) { return true }
+            }
+        } catch {
+            return false
+        }
+
+        return false
+    }
+
+    func createWalletBackup() throws {
+        guard let backupFolder = TariLib.shared.tariWallet?.publicKey.0?.hex.0 else { throw ICloudBackupWalletError.iCloudContainerNotFound }
+
+        let fileURL = try zipBackupFiles()
```

Comment:
We're really after a single file now, so you can be specific and go for `zipWalletDatabase()`.